src/java.base/share/classes/java/lang/Integer.java

```diff
@@ -453,16 +454,104 @@ private static void formatUnsignedIntUTF16(int val, int shift, byte[] buf, int l
      */
     @IntrinsicCandidate
     public static String toString(int i) {
-        int size = stringSize(i);
-        if (COMPACT_STRINGS) {
-            byte[] buf = new byte[size];
-            getChars(i, size, buf);
-            return new String(buf, LATIN1);
-        } else {
+        if (!COMPACT_STRINGS) {
+            int size = stringSize(i);
             byte[] buf = new byte[size * 2];
             StringUTF16.getChars(i, size, buf);
             return new String(buf, UTF16);
         }
+
+        if (i == Integer.MIN_VALUE) {
+            return "-2147483648";
+        }
+
+        boolean negative = i < 0;
+        if (negative) {
+            i = -i;
+        }
+        final int[] digits = DecimalDigits.DIGITS;
+
+        int off = 0;
+        final int q1 = i / 1000;
+
+        byte[] buf;
+        if (q1 == 0) {
```

Comment:
You could postpone the division to compute `q1` later, and here test whether `i < 1_000`.
I've no statistics, but I guess it makes sense to optimize for tiny `int`s by avoiding a division.

Reply:
I have submitted the changes and updated the benchmark results, which shows an improvement.